doc/control-mechanisms/reporting.md

```diff
@@ -0,0 +1,81 @@
+---
+title: "Reporting Konzept"
+---
+
+## Ziel
+
+Das Ziel dieses Dokuments ist es, die Konzepte zum **Reporting** zentral zu sammeln.
+**Reporting** bezeichnet den Prozess des Erfassens, Bewertens und Speicherns einer Antwort auf
+eine Frage im Dungeon-Kontext.
+
+## Beteiligte Komponenten
+
+**Task-Definition**
+
+Die Task-Definition definiert die Aufgabe, Beispiel:
+
+```
+single_choice_task my_task {
+  answers: ["a", "b", "c"],  // die Antwortmöglichkeiten
+  correct_answer: answers.0, // die korrekte Antwortmöglichkeit
+  fn_score: score            // die Scoring-Funktion/Bewertungsfunktion
+}
+```
+
+**Scoring-Funktion**
+
+Die **Scoring-Funktion** implementiert die Art, wie eine Aufgabe bewertet wird.
+Eine beispielhafte DSL-Implementierung könnte so aussehen:
+
+```
+fn score(single_choice_task t, string a) -> float {
+  if t.correct_answer == a
+    return 1.0;
+  else
+    return 0.0;
+}
+```
+
+Die **Scoring-Funktion** nimmt zwei Parameter entgegen: das Aufgabenobjekt (was die **Task-Definition**
+repräsentiert), und die gegebenen Antwort (hier als `string`).
+Sie gibt einen Fließkommawert zurück, der die Bewertung der Antwort repräsentiert: 1.0 = richtig, 0.0 = falsch.
+
+**Szenario-Definition**
+
+Die **Szenario-Definition** ist dafür zuständig, die Verbindung zwischen der abstrakten Aufgaben-Definition und
+der konkreten Umsetzung im Dungeon herzustellen. Hierzu zählt auch Verknüpfung von Event-Handler Funktionen
+mit Events im Kontext von Entitäten, um auf bestimmte Ereignisse im Dungeon zu reagieren.
+Da die "Abgabe" einer Aufgabe im Dungeon-Kontext durch viele szenario-abhängige Ereignisse abgebildet werden kann,
+muss auch der Aufruf der **Scoring-Funktion** und weiterer Reporting-Logik in die Szenario-Definition
+integriert werden.
+
+```
+fn handler_function(entity e) {
+  // andere Event-Handler Logik
+
+  report(e.properties.task, e.properties.task_answer_value)
```

Comment:
Hmmm. Das würde bedeuten, dass in einer `Entity` auch die Tasks gespeichert werden? @AMatutat

Reply:
Oh, steht ganz unten. 

@AMatutat Hast Du das auf dem Schirm?

Reply:
Die `properties` sind aktuell `components`. Man bräuchte dann wohl noch eine `TaskComponent`, die die oben definierten Sachen (Fragen, Anworten, korrekte Antwort, Scoring) enthält.

@malt-r Sollte das hier auch gleich so umgesetzt werden oder ist es besser, das intern im Interpreter in die richtigen Strukturen zu übersetzen?

Reply:
> @AMatutat Hast Du das auf dem Schirm?

Denn Callback kram? Ja. Dafür haben wir "flächendeckend" das Strategy-Pattern implementiert. 
Die Idee ist, dass der DSL-Callback dann z.B eine "onInteraction"-Strategie sein wird. 
Und die Menge der in der DSL formulierten Entitäten + Callbacks ergibt dann den Task. (@malt-r das war vermutlich etwas stark vereinfacht, oder?)

Reply:
@AMatutat Ein dediziertes `TaskComponent`, bzw. besser `TaskReferenceComponent` ist sinnvoll, das habe ich auch schon häufiger mal angesprochen. So würde man über die `Entity` direkt an die `task`-Definition kommen. Hab dafür eine Notiz im Dokument ergänzt. Ich bin mir gerade nicht sicher, ob es dafür schon ein Issue gibt.

Reply:
Steht jetzt hier: https://github.com/Programmiermethoden/Dungeon/issues/685